hazelcast-jet-core/src/main/java/com/hazelcast/jet/pipeline/test/LongStreamSourceP.java

```diff
@@ -0,0 +1,133 @@
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+
+import javax.annotation.Nonnull;
+
+import static com.hazelcast.jet.impl.JetEvent.jetEvent;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Implements the {@link TestSources#longStream} source.
+ *
+ * @since 4.3
+ */
+public class LongStreamSourceP extends AbstractProcessor {
+
+    private static final long SOURCE_THROUGHPUT_REPORTING_PERIOD_SECONDS = 10;
+
+    private static final long REPORT_PERIOD_NANOS = SECONDS.toNanos(SOURCE_THROUGHPUT_REPORTING_PERIOD_SECONDS);
+    private static final long HICCUP_REPORT_THRESHOLD_MILLIS = 10;
+    private final long nanoTimeMillisToCurrentTimeMillis = determineTimeOffset();
+    private final long startTime;
+    private final long itemsPerSecond;
+    private final ILogger logger = Logger.getLogger(LongStreamSourceP.class);
+    private final long wmGranularity;
+    private final long wmOffset;
+    private long globalProcessorIndex;
+    private long totalParallelism;
+    private long emitPeriod;
+
+    private final AppendableTraverser<Object> traverser = new AppendableTraverser<>(2);
+    private long emitSchedule;
+    private long lastReport;
+    private long counterAtLastReport;
+    private long lastCallNanos;
+    private long counter;
+    private long lastEmittedWm;
+    private long nowNanos;
+
+    LongStreamSourceP(
+            long startTime,
+            long itemsPerSecond,
+            EventTimePolicy<? super Long> eventTimePolicy
+    ) {
+        this.wmGranularity = eventTimePolicy.watermarkThrottlingFrameSize();
+        this.wmOffset = eventTimePolicy.watermarkThrottlingFrameOffset();
+        this.startTime = MILLISECONDS.toNanos(startTime + nanoTimeMillisToCurrentTimeMillis);
+        this.itemsPerSecond = itemsPerSecond;
+    }
+
+    @Override
+    protected void init(@Nonnull Context context) {
+        totalParallelism = context.totalParallelism();
+        globalProcessorIndex = context.globalProcessorIndex();
+        emitPeriod = SECONDS.toNanos(1) * totalParallelism / itemsPerSecond;
+        lastCallNanos = lastReport = emitSchedule =
+                startTime + SECONDS.toNanos(1) * globalProcessorIndex / itemsPerSecond;
+    }
+
+    @Override
+    public boolean complete() {
+        nowNanos = System.nanoTime();
+        emitEvents();
+        detectAndReportHiccup();
+        if (logger.isFineEnabled()) {
+            reportThroughput();
+        }
+        return false;
+    }
+
+    private void emitEvents() {
+        while (emitFromTraverser(traverser) && emitSchedule <= nowNanos) {
+            long timestamp = NANOSECONDS.toMillis(emitSchedule) - nanoTimeMillisToCurrentTimeMillis;
+            traverser.append(jetEvent(timestamp, counter * totalParallelism + globalProcessorIndex));
+            counter++;
```

Comment:
This would break the invariant we rely on elsewhere, for the `counter` value to match the number of emitted events. The overhead of calculation isn't a concern, I think.